bundles/binding/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/ADZoneTracker.java

```diff
@@ -0,0 +1,245 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Track zone faults / restores by zone ID watching the alarm panel
+ * message stream
+ *
+ * @author Sean Mathews <coder@f34r.com>
+ * @since 1.6.0
+ */
+public class ADZoneTracker {
+
+    private static final Logger logger = LoggerFactory.getLogger(ADZoneTracker.class);
+
+    private ArrayList<ADZone> _zoneList;
+    private boolean _lastReady;
+    private int _lastZone = 0;
+
+    // TODO: zone timeout handling. zone list garbage collector.
+    long _currTime = 0;
+
+
+    ADZoneTracker() {
+        _zoneList = new ArrayList<ADZone>();
+        _lastReady = false;
+        _currTime = System.nanoTime();
+    }
+
+    /**
+     * get the state of a zone
+     *
+     * @param Zone - the zone
+     * @return return zone status or null if the zone is invalid
+     *     -1 = unknown
+     *      0 = restored
+     *      1 = faulted
+     */
+    public boolean getZoneState(int Zone) {
+
+        /** find the zone in our List */
+        ADZone f = new ADZone(Zone);
+        int z = _zoneList.indexOf(f);
+
+        if ( z > -1 ) {
+            f = _zoneList.get(z);
+        }
+
+        return f.getState();
+    }
+
+    /**
+     * set the state of a zone
+     *
+     * @param Zone  - the zone
+     * @param State - faulted or not boolean
+     * @return zone status
+     *     -1 = unknown
+     *      0 = restored
+     *      1 = faulted
+     */
+    public boolean setZoneState(int Zone, boolean State) {
+
+        /** find the zone in our List */
+        ADZone f = new ADZone(Zone);
+        int z = _zoneList.indexOf(f);
+
+        if ( z > -1 ) {
+            f.setState(State);
+            f = _zoneList.set(z,f);
+        } else {
+            f.setState(State);
+            _zoneList.add(f);
+        }
+
+        return f.getState();
+    }
+
+    /**
+     * Clear all tracked zones happens when we see a "READY" from the panel
+     *
+     */
+     public void clearAll() {
+         _zoneList.clear();
+         _lastZone = 0;
+         _lastReady = true;
+     }
+
+    /**
+     * Update a zone state from messages from panel. Must be in the
+     * correct sequence and not modified from what the panel provides
+     * to properly track zones
+     *
+     * @param Zone  - the zone
+     * @param State - the state
+     * @return ArrayList of zones effected
+     */
+     public ArrayList updateZone(int Zone, boolean State) {
+
+        /** return a list of update zones */
+        ArrayList<Integer> updateList =  new ArrayList<Integer>();
+
+        /** add this zone */
+        updateList.add(Zone);
+
+        /** our list iterator */
+        ListIterator litr = _zoneList.listIterator();
+
+        /** zone exists so check for any pruning */
+        if(_zoneList.contains(new ADZone(Zone))) {
+            /** go to our last zone updated */
+            ADZone f = new ADZone(_lastZone);
+            int z = _zoneList.indexOf(f);
+
+		    boolean bFoundLast = false;
```

Comment:
I use MonoDevelop most of the time for C or C# or I just use VI as I mostly do embedded C code. I have not installed Eclipse since my last system(s) rebuild. at least for me in the past not sure if its better now the IDE layout was confusing and picky about where files go. Maybe someday I will play again but I dont do much Java in the past few years.

Thanks for the Tip! will have to find some time to play.